mavros_extras/src/plugins/play_tune.cpp

```diff
@@ -0,0 +1,59 @@
+#include <mavros/mavros_plugin.h>
+#include <mavros_msgs/PlayTuneV2.h>
+#include <cstring>
+
+namespace mavros
+{
+namespace extra_plugins
+{
+
+class PlayTunePlugin : public plugin::PluginBase
+{
+  public:
+    PlayTunePlugin() : PluginBase(), nh("~") {}
+
+    void initialize(UAS& uas_) override
+    {
+        PluginBase::initialize(uas_);
+        sub = nh.subscribe("play_tune", 1, &PlayTunePlugin::callback, this);
+    }
+
+    Subscriptions get_subscriptions() override { return { /* No subscriptions */ }; }
+
+  private:
+    ros::NodeHandle nh;
+    ros::Subscriber sub;
+
+    void callback(const mavros_msgs::PlayTuneV2::ConstPtr& tune)
+    {
+        auto msg = mavlink::common::msg::PLAY_TUNE_V2{};
+        msg.target_system = m_uas->get_tgt_system();
+        msg.target_component = m_uas->get_tgt_component();
+
+        switch (tune->format)
+        {
+            case mavros_msgs::PlayTuneV2::TUNE_FORMAT_QBASIC1_1:
+                msg.format = static_cast<uint32_t>(mavlink::common::TUNE_FORMAT::QBASIC1_1);
+                break;
+
+            case mavros_msgs::PlayTuneV2::TUNE_FORMAT_MML_MODERN:
+                msg.format = static_cast<uint32_t>(mavlink::common::TUNE_FORMAT::MML_MODERN);
+                break;
+
+            default:
+                ROS_ERROR_NAMED("play_tune", "Invalid tune format: '%i'", tune->format);
+                return;
+        }
+
+        std::strncpy(msg.tune.data(), tune->tune.c_str(), sizeof(msg.tune));
+        msg.tune.back() = '\0';
```

Comment:
Please use `mavlink::set_string_z()`.

https://github.com/ArduPilot/pymavlink/blob/b1efa7e5fd50561dfa9f8a491e4a55c8e0eedec9/generator/CPP11/include_v2.0/message.hpp#L143

Reply:
Thanks, that's much better. `strncpy` works, but it sure feels bug prone.